site: custom index docs (#618)
kubernetes-sigs/krew: site/content/docs/user-guide/using-custom-indexes.md

---
title: Using Custom Plugin Indexes
slug: custom-indexes
weight: 800
draft: true
---

Plugin indexes contain plugin manifests, which are documents that describe
installation procedure for a plugin. For discovery purposes, Krew comes with a
`default` plugin index,  plugins hosted in the [`krew-index`
repository](https://github.com/kubernetes-sigs/krew-index).

However, some plugin authors may choose to host their own indexes that contain
their curation of kubectl plugins. These are called "custom plugin indexes".

> ## Adding a custom index

Reviewer: Can you also add an "advanced" section, where you explain how users/vendors could substitute the krew default index with their own?

Author: Whoops I missed this comment when I was looking at your review yesterday. Sure I can add this

A custom plugin index can be added with the `kubectl krew index add` command:

```sh
{{<prompt>}}kubectl krew index add foo https://github.com/foo/custom-index.git
```

The URI you use can be any [git remote](https://git-scm.com/docs/git-remote)
(e.g., `git@github.com:foo/custom-index.git`).

## Removing a custom index

You can remove a custom plugin index by passing the name it was added with to
the remove command:

```sh
{{<prompt>}}kubectl krew index remove foo
```

## Listing indexes

> To see what indexes you have added you can run the list command

Reviewer: ```suggestion
To see what indexes you have added run the list command
```

```sh
{{<prompt>}}kubectl krew index list
{{<output>}}INDEX    URL
default  https://github.com/kubernetes-sigs/krew-index.git
foo      https://github.com/foo/custom-index.git{{</output>}}
```

## Installing plugins from custom indexes

Commands for managing plugins (e.g. `install`, `upgrade`) work with custom
indexes as well.

> By default, Krew prefixes plugins with a `default/` prefix. However, if you have
> a plugin installed from a custom index, you need to specify it in format

Reviewer: ```suggestion
By default, Krew prefixes plugins with a `default/` prefix. So if
a plugin comes from a custom index, you need to specify it in format
```

`INDEX_NAME/PLUGIN_NAME`.

For example, to install a plugin named `bar` from custom index `foo`:

```sh
{{<prompt>}}kubectl krew install foo/bar
```

Similarly:

- To list all plugins (including the ones from custom indexes), run:

    ```sh
    {{<prompt>}}kubectl krew search
    ```

- To remove a plugin, you don't need to specify its index:

    ```sh
    {{<prompt>}}kubectl krew uninstall PLUGIN_NAME
    ```

- To get information about a plugin from a custom index:

    ```sh

>     {{<prompt>}}kubectl krew info INDEX_NAME/PLUGIN_NAME

Reviewer: Is that also correct if the plugin is already installed? I don't remember..

Author: Yes since they can have name conflicts, we did it this way so that you can get info for plugin `foo` in default and custom index.

---
title: Using Custom Plugin Indexes
slug: custom-indexes
weight: 800
draft: true
---

Plugin indexes contain plugin manifests, which are documents that describe
installation procedure for a plugin. For discovery purposes, Krew comes with a
`default` plugin index, plugins hosted in the [`krew-index`
repository](https://github.com/kubernetes-sigs/krew-index).

However, some plugin authors may choose to host their own indexes that contain
their curation of kubectl plugins. These are called "custom plugin indexes".

> ## Adding a custom index

Reviewer 2: I think the title isn’t good. 
It would be better to explain in this section “The default index” and

* how it can be removed
* or pointed to another repo
* and how plugins without an explicit index name mean “default”

Author: I feel like the adding/removing/listing indexes sections are important to showcase the new index command right? I agree though that a section explaining the default index would be good, I can move the advanced usage section I had and change it to reflect the points you listed.

Reviewer: Yes, I think this should work well. You could also explain how plugin names are resolved. Iow, that a `default/` prefix is added implicitly, if no index name is given. Then you could also explain how plugin name clashes are handled and remove the `Duplicate plugin names` section in the other page.

weight: 800
draft: true
---

Plugin indexes contain plugin manifests, which are documents that describe
installation procedure for a plugin. For discovery purposes, Krew comes with a
`default` plugin index, plugins hosted in the [`krew-index` repository][ki].

However, some plugin authors may choose to host their own indexes that contain
their curation of kubectl plugins. These are called "custom plugin indexes".

## Adding a custom index

A custom plugin index can be added with the `kubectl krew index add` command:

```sh
{{<prompt>}}kubectl krew index add foo https://github.com/foo/custom-index.git
```

The URI you use can be any [git remote](https://git-scm.com/docs/git-remote)
(e.g., `git@github.com:foo/custom-index.git`).

## Removing a custom index

You can remove a custom plugin index by passing the name it was added with to
the remove command:

```sh
{{<prompt>}}kubectl krew index remove foo
```

## Listing indexes

To see what indexes you have added run the list command:

```sh
{{<prompt>}}kubectl krew index list
{{<output>}}INDEX    URL
default  https://github.com/kubernetes-sigs/krew-index.git
foo      https://github.com/foo/custom-index.git{{</output>}}
```

> ## The default index

Reviewer 2: Can we perhaps use code block notation for all “default”s in this paragraph?

> ## The default index

Reviewer 2: This section should ideally come **after** installing plugins from custom indexes (keep in mind most people might not care about the default index nuances)

> When a plugin doesn't have an explicit `INDEX_NAME` prefix it refers to a plugin

Reviewer 2: , it

from the default index. These plugins have an implicit `default/` prepended to
them in Krew commands. The `INDEX_NAME` prefix is used to differentiate plugins
with the same name across different indexes.

Krew ships with [`krew-index`][ki] as the default index, but this can be removed

> by passing `default` to the remove command. Once this is removed, you can add

Reviewer 2: > ...Removed using the “kubectl krew index remove default” command
